In [47]:
#import validaciones
import validaciones2

# Ejercicios Python

## Zen of Python

Lee una vez mas el Zen of Python antes de iniciar con la tarea

In [48]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## 1 - XOR

En la siguiente celda crea una funcion para la compuerta logica XOR 

<img src="assets/xor.png" width="400">


In [49]:
def xor(a,b):
    """
    Esta funcion recibe dos valores booleanos, ejemplo xor(1, 0) o True y False
    y retorna 0 o 1 como resultado de la operacion logica XOR
    """
    if a == 1:
        if b == 1:
            return 0
        else: 
            return 1
        
    else :
        if b == 1:
            return 1
        else :
            return 0
        
    

In [50]:
xor(0 , 1) 

1

In [51]:
xor(0 , 0)

0

In [52]:
xor(True , False)

1

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

In [53]:
#validaciones.xor_function(xor)
validaciones2.xor_function(xor)

Prueba 1
Prueba 1 [CORRECTA]
Prueba 2
Prueba 2 [CORRECTA]


## 2 - Sumando N series

Tenemos una secuencia cuyo $n^{th}$ es:
\begin{equation*}
\ T_n = n^2 - (n-1)^2
\end{equation*}

Hay que evaluar las series:
\begin{equation*}
\ S_n = T_1 + T_2 + ... + T_n
\end{equation*}


**Instrucciones**
En la siguiente celda en la funcion summingSeries debemos **retornar** en valor de $S_n$, tenemos como entrada n

**Ejemplo 1:**

Si la entrada es 2

Salida:
4

Explicacion:

$T_1 = 1$

$T_2 = 3$

$S_2 = T_1 + T_2$

$S_2 = 4 $



**Tip :**
Antes te implementar $S_n$ la solución, analicemos algebraicamente el problema para encontrar una solución equivalente optima.


In [54]:
def summatoriaSerie(n):
    """
    La funcion recibe como parametro n y retorna el valor se la serie
    """
    return n**2

print(summatoriaSerie(2))

4


#### 2.1 - Ciclos

Crea un ciclo for que imprima el resultado de la funcion **summingSeries** desde 1 hasta 10

In [55]:
for i in range(1,11):
    print(summatoriaSerie(i))

1
4
9
16
25
36
49
64
81
100


### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [56]:
#validaciones.summingSeries_function(summatoriaSerie)
validaciones2.summingSeries_function(summatoriaSerie)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]


## 3 - Factores Primos

Para cada $n$ queremos obtener el conteo maximo de numeros primos unicos en en rango inclusivo $[1, n]$,
y retornar el valor del conteo en una nueva linea

**Nota:** Recuerda que un numero primo solo es divisible por el mismo y que 1 no es un numero primo

##### Ejemplos:

###### Ejemplo 1
Entrada: 1

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,1]$ es $0$, porque $1$ no es un numero primo.

###### Ejemplo 2
Entrada: 3

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,3]$ es $1$, porque el numero $3$ tiene 1 factor primo unico (el mismo) 

###### Ejemplo 3
Entrada: 500

Salida esperada: 4

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,500]$ es $4$, porque el producto de los primeros cuatro numeros primos unicos es $2 \times 3 \times 5 \times 7 = 210$ y no hay ningun numero primo unico que multiplicado por el resultado sea $	\leqslant 500$



#### Tip: 
Utiliza la funcion `range()` de python


In [57]:
def isPrime(n):
    if n > 1:
        for i in range(2, n):
            if (n % i) == 0:
                return False
        else:
            return True
    else:
        False

def conteoPrimos(n):
    """
    Esta funcion recibe como parametro n y 
    retorna el conteo maximo de numeros primos unicos
    """
    conteo = 0
    factor = 1
    for i in range(1, n + 1):
        # print(i)
        if isPrime(i):
            factor *= i
            if(factor <= n):
                conteo += 1
            else:
                break
    
    return conteo

print(conteoPrimos(1))
print(conteoPrimos(500))

0
4


### Pruebas

In [58]:
conteoPrimos(1)

0

In [59]:
conteoPrimos(500)

4

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [60]:
#validaciones.primeCount_function(conteoPrimos)
validaciones2.primeCount_function(conteoPrimos)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]
Prueba 7
Prueba 7[CORRECTA]
Prueba 8
Prueba 8[CORRECTA]
Prueba 9
Prueba 9[CORRECTA]
Prueba 10
Prueba 10[CORRECTA]
Prueba 11
Prueba 11[CORRECTA]
Prueba 12
Prueba 12[CORRECTA]


## 4 - Metodo de Euler

El metodo de Euler se utiliza para aproximar la solucion particular de una ecuacion diferencial dado un valor inicial. Con esta informacion se sabe que la grafica de esa solucion pasa a traves del punto $(x_0,y_0)$ y tiene una pendiente $F(x_0,y_0)$ en ese punto, esto da un punto de partida para aproximar la solucion.

A partir del punto inicial, se sigue en la direccion indicada por la pendiente. Mediante un pequeño paso $h$, se mueve a lo largo de la recta tangente hasta llegar al punto $(x_1,y_1)$ donde:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ y_1 = y_0 + hF(x_0,y_0)\\
\end{equation*}


<img src="assets/Euler.png" width="350">

Como se muestra en la figura se considera $(x_1,y_1)$ como un nuevo punto inicial.

Los valores de $x$ son:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ x_2 = x_1 + h       \\
\    .                \\
\    .                \\
\    .                \\
\ x_n = x_{n-1} + h   \\
\end{equation*}


Los valores de $y$ son:
\begin{equation*}
\ y_1 = y_0 + hF(x_0,y_0)\\
\ y_1 = y_1 + hF(x_1,y_1)\\
\       . \\
\       . \\
\       . \\
\ y_n = y_{n-1} + hF(x_{n-1},y_{n-1})\\
\end{equation*}


#### Problema 4
El ejecicio consiste en implementar la funcion *euler_metod* esta funcion recibe como parametro el 
* punto inicial $x_0$
* punto inicial $y_0$
* Paso $h$ 
* numero de iteraciones $n$ 
* La equacion diferencial que queremos aproximar, esta funcion tambien recibe los parametros $x$ y $y$
* La funcion de la solucion exacta que tambien tambien recibe los parametros $x$ y $y$

La funcion debe retornar el valor de $y_n$ y el error
$error=|y - y_n|$ 

In [68]:
import math 

In [72]:
def y_prime(x, y):
    return -2*y

def y(x):
    return 4 * math.exp(-2*x)

def euler_method(x,y,h,n, diff_eq, func):
    """
    Parametros x, y punto inicial, logitud del paso h , n numero de iteraciones,
    diff_eq la equacion diferencial y func la funcion de la solucion exacta
    Retorna y y el error
    """
    previous_y = y
    previous_x = x + h
    for i in range(0, n):
        print("x: %f" % previous_x)
        exact_y = func(previous_x)
        previous_y = previous_y + h*(diff_eq(previous_x,previous_y))
        previous_x = previous_x + h
        error = abs(exact_y - previous_y)
        print("y_%d: %f" % (i,previous_y))
        print("y: %f" % exact_y)
        print('error: %f' % error)
        print('*****************************************************')
        
euler_method(0,4,0.1,10, y_prime,y)
        

x: 0.100000
y_0: 3.200000
y: 3.274923
error: 0.074923
*****************************************************
x: 0.200000
y_1: 2.560000
y: 2.681280
error: 0.121280
*****************************************************
x: 0.300000
y_2: 2.048000
y: 2.195247
error: 0.147247
*****************************************************
x: 0.400000
y_3: 1.638400
y: 1.797316
error: 0.158916
*****************************************************
x: 0.500000
y_4: 1.310720
y: 1.471518
error: 0.160798
*****************************************************
x: 0.600000
y_5: 1.048576
y: 1.204777
error: 0.156201
*****************************************************
x: 0.700000
y_6: 0.838861
y: 0.986388
error: 0.147527
*****************************************************
x: 0.800000
y_7: 0.671089
y: 0.807586
error: 0.136497
*****************************************************
x: 0.900000
y_8: 0.536871
y: 0.661196
error: 0.124325
*****************************************************
x: 1.000000
y_9: 0.429497
y:

### Ejemplo

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial y}{\partial x} = -2y \\
\end{equation*}
Donde $y(0) = 4$ vamos a unsa un de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ y = 4e^{-2x} \\
\end{equation*}

In [70]:
### Implementando las funciones del ejemplo
def y_prime(x, y):
    return -2*y

def y(x):
    return 4 * math.exp(-2*x)

In [73]:
"""
Llamamos a la funcion de euler
euler_method(0,4,0.1,10, y_prime,y)
Parametros :
x = 0
y = 4
d = 0.1
n = 10
diff_eq = y_prime
func = y
"""

euler_method(0,4,0.1,10, y_prime,y)

x: 0.100000
y_0: 3.200000
y: 3.274923
error: 0.074923
*****************************************************
x: 0.200000
y_1: 2.560000
y: 2.681280
error: 0.121280
*****************************************************
x: 0.300000
y_2: 2.048000
y: 2.195247
error: 0.147247
*****************************************************
x: 0.400000
y_3: 1.638400
y: 1.797316
error: 0.158916
*****************************************************
x: 0.500000
y_4: 1.310720
y: 1.471518
error: 0.160798
*****************************************************
x: 0.600000
y_5: 1.048576
y: 1.204777
error: 0.156201
*****************************************************
x: 0.700000
y_6: 0.838861
y: 0.986388
error: 0.147527
*****************************************************
x: 0.800000
y_7: 0.671089
y: 0.807586
error: 0.136497
*****************************************************
x: 0.900000
y_8: 0.536871
y: 0.661196
error: 0.124325
*****************************************************
x: 1.000000
y_9: 0.429497
y:

### Solucion:

Utiliza esta tabla como guia para validar que tu funcion este correcta,
y reemplaza los valores que faltan con la salida de tu funcion.

| x     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| y     | 4 | 3.2749      | 2.6813 | 1.7973 | 1.2048 | 0.8076 | 0.5413       |
| $y_n$ | 4 | 3.200000    | 2.56   | 1.6384 | 1.0486 | 0.6711 | 0.429497     |
| error | 0 | 0.074923    | 0.1213 | 0.1589 | 0.1562 | 0.1365 | 0.111844     |

### 4.1 Probando la implementacion con una aplicación real de data science

Podemos utilizar ecuaciones diferenciales  como modelos matemáticos de fenónemos reales, por ejemplo si llegas a trabajar como data scientist para una farmaceutica  puede que interese analizar la desintegración radioactiva de cierta sustancia y como esto cambia en el tiempo
#### Desintegracion Radiactiva

$A$ es la cantidad de sustancia en el tiempo $t$, la derivada $\frac{\partial A}{\partial t}$ es rapidez a la que se desintegran los nucleos de una sustancia.

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial A}{\partial t} =  -0.56A \\
\end{equation*}
Donde $A(0) = 1$ vamos a un tamaño de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ A(t) = e^{-0.56t} \\
\end{equation*}

**Nota** En este caso sencillo, poseemos la solución exacta a la ED por lo cual no se necesita una solución aproximada , pero en aplicaciones y proyectos reales muchas veces la solución exacta no se posee, o es muy difícil de calcular y recurrimos a soluciones aproximadas por computadora , para este ejercicio el saber la solución exacta nos sirve para medir la exactitud de nuestra solución aproximada.

## 5 - Rotacion de listas hacia la izquierda

Una rotacion a la izquierda en una lista mueve cada elemento de la lista una vez hacia la izquierda. Por ejemplo, si hacemos $1$ rotacion en en el arreglo $[1,2,3,4,5]$, el arreglo resultante seria $[2,3,4,5,1]$

#### Descipcion de la funcion
Dado un arreglo $a$ de $n$ integers y un numero $d$ de rotaciones, realizar $d$ rotaciones hacia la izquierda. La funcion debe retorner el arreglo re-ordenado.

**rotIzquierda** tiene los siguientes parametros:

* $a$ Un arreglo de numeros entreros.
* $d$ numero de rotaciones.

In [74]:
def rotIzquierda(a, d):
    # tu codigo aqui
    arr_len = len(a)
    if d > arr_len:
        d = d % arr_len
        
    result = [None] * (arr_len)
    for i in range(0, arr_len):
        index = i - d
        result[index] = a[i]
    
    return result

In [75]:
rotIzquierda([1, 2, 3, 4, 5], 4)
# Resultado: [5, 1, 2, 3, 4]

[5, 1, 2, 3, 4]

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

#### TIP:
Utiliza **slice** en el array para optimizar la funcion

In [76]:
#validaciones.rotIzquierda_function(rotIzquierda)
validaciones2.rotIzquierda_function(rotIzquierda)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
